Motorway.cpp

```diff
@@ -0,0 +1,61 @@
+
+#include "Motorway.h"
+
+void Motorway::addCar(Car* car)
+{
+    carsArray.push_back(*car);
+}
+
+void Motorway::addMotorBike(Motorbike* motorbike)
+{
+    motorbikesArray.push_back(*motorbike);
+}
+
+void Motorway::addLorry(Lorry* lorry)
+{
+    lorriesArray.push_back(*lorry);
+}
+
```

Comment:
I hope you realize all of these are making copies of the pointed-to object!

Reply:
Interesting.  So I changed the vectors so that they contain pointers and not objects and that seems to have stopped unnecessary temporary objects being created and destroyed.  The way I debugged was adding std::out statements to the constructors and destructors.  But I still have a few questions:

1) In the previous code where the vectors didn't contain pointers, it seemed only the destructors of the car/motorbike/lorry objects were being called an extra time, but if a copy was being made why wouldn't both the constructor and destructor be called for each temporary copy?

2) I tried using reserve and emplace but that didn't seem to make any difference.  When would it?

3) I tried to use references but everything blew up.  Is there a way of achieving the same result, that is no copying, without using pointers?

Thanks

Reply:
1) it's calling the implicitly declared Copy Constructor
2) move semantics would prevent object duplication. but you're not there yet.
3) https://en.cppreference.com/w/cpp/utility/functional/reference_wrapper.  
`vector<T*>` is the best approach as long as you follow the rules for using pointers.